### PR TITLE
Register watchable build directories before checking the configuration cache fingerprint

### PR DESCRIPTION
**Pull request by bamboo**
To avoid `Found existing snapshot at ... for unwatched hierarchy ...` exceptions when registering included builds after checking build script file fingerprints.

### CODE REVIEW

**Comment by bamboo** [comment]
@bot-gradle test this

**Comment by bamboo** [comment]
@bot-gradle test RFN

**Comment by bamboo** [comment]
@bot-gradle test QF

**Comment by bamboo** [comment]
@bot-gradle test this

**Comment by wolfs** [comment]
I don't understand the change, yet. 

Can you give me a little more context? It seems that when loading from the configuration we use `FileSystemAccess` before adding the project root directory as a watchable directory. Though I don't know where. And now we load the watched directories from the configuration cache before using `FileSystemAccess`?

**Comment by bamboo** [comment]
> Can you give me a little more context? It seems that when loading from the configuration we use `FileSystemAccess` before adding the project root directory as a watchable directory. Though I don't know where. And now we load the watched directories from the configuration cache before using `FileSystemAccess`?

Exactly.

Gradle checks the fingerprint of the configuration cache before creating any builds and projects. Checking the fingerprint might require `FileSystemAccess` in order to [compute the hash for a build script included in the fingerprint](https://github.com/gradle/gradle/blob/1d53f7c952bec75163a9759f55f880b9d6fa60d3/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintController.kt#L185) or `FileCollectionFingerprinter` to [get the fingerprint for some build logic task inputs](https://github.com/gradle/gradle/blob/1d53f7c952bec75163a9759f55f880b9d6fa60d3/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintController.kt#L188).

The full fingerprint checking process is implemented in
https://github.com/gradle/gradle/blob/83761d0b835e5977ab66cb8c1dba34f91a8dd3f1/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintChecker.kt#L47-L93

Before this change, the fingerprint check could access a build script file from an included build before the root directory of that build being registered as a watchable hierarchy.

This PR records the set of included build directories in the fingerprint header so they can all be registered as watchable hierarchies before the checking process begins.

**Comment by bamboo** [comment]
@wolfs, I'm merging it now so I don't have to rebase it for release Tomorrow but please do provide feedback.

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt [hunk 1]
```diff
@@ -82,11 +82,14 @@ class ConfigurationCacheState(
...
-        writeRootBuild(build)
-        writeInt(0x1ecac8e)
-    }
+    /**
+     * Writes the state for the whole build starting from the given root [build] and returns the set
+     * of stored included build directories.
+     */
+    suspend fun DefaultWriteContext.writeRootBuildState(build: VintageGradleBuild): HashSet<File> =
+        writeRootBuild(build).also {
+            writeInt(0x1ecac8e)
```
Comment (wolfs) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt:91]: What is this magic marker?

**Review by wolfs — COMMENTED** [review]
LGTM! I left a question.

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt (hunk 1 above)
Comment (bamboo) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt:91]: Just a sanity check. We had a bug in the past where the deserialization code was reading less work graph elements than written so I added this marker to ensure defects like that don't go unnoticed.

**Comment by bamboo** [comment]
Thanks for the review, @wolfs!

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt (hunk 1 above)
Comment (wolfs) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt:91]: Can you add a comment here so this information is conserved for future readers?

Comment (bamboo) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt:91]: Will do!

**Pull request merged**